engines/router/missionctl/server/upi/server.go

```diff
@@ -0,0 +1,176 @@
+package upi
+
+import (
+	"context"
+	"net"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/caraml-dev/turing/engines/router/missionctl"
+	"github.com/caraml-dev/turing/engines/router/missionctl/errors"
+	"github.com/caraml-dev/turing/engines/router/missionctl/instrumentation/metrics"
+	"github.com/caraml-dev/turing/engines/router/missionctl/instrumentation/tracing"
+	"github.com/caraml-dev/turing/engines/router/missionctl/log"
+	"github.com/caraml-dev/turing/engines/router/missionctl/log/resultlog"
+	"github.com/caraml-dev/turing/engines/router/missionctl/server/constant"
+	"github.com/caraml-dev/turing/engines/router/missionctl/turingctx"
+	upiv1 "github.com/caraml-dev/universal-prediction-interface/gen/go/grpc/caraml/upi/v1"
+	fiberGrpc "github.com/gojek/fiber/grpc"
+	fiberProtocol "github.com/gojek/fiber/protocol"
+	"github.com/opentracing/opentracing-go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/protobuf/proto"
+)
+
+const tracingComponentID = "grpc_handler"
+
+type Server struct {
+	upiv1.UnimplementedUniversalPredictionServiceServer
+
+	missionControl missionctl.MissionControlUPI
+}
+
+func NewUPIServer(mc missionctl.MissionControlUPI) *Server {
+	return &Server{
+		missionControl: mc,
+	}
+}
+
+func (us *Server) Run(listener net.Listener) {
+	s := grpc.NewServer()
+	//TODO: the unmarshalling can be done more efficiently by using partial deserialization
+	upiv1.RegisterUniversalPredictionServiceServer(s, us)
+	reflection.Register(s)
+
+	errChan := make(chan error, 1)
+	stopChan := make(chan os.Signal, 1)
+
+	// bind OS events to the signal channel
+	signal.Notify(stopChan, syscall.SIGTERM, syscall.SIGINT)
+
+	go func() {
+		if err := s.Serve(listener); err != nil {
+			errChan <- err
+		}
+	}()
+
+	defer func() {
+		s.GracefulStop()
+	}()
+
+	// block until either OS signal, or server fatal error
+	select {
+	case err := <-errChan:
+		log.Glob().Errorf("Failed to start Turing Mission Control API: %s", err)
+	case <-stopChan:
+		log.Glob().Info("Signal to stop server")
+	}
+
+}
+
+func (us *Server) PredictValues(ctx context.Context, req *upiv1.PredictValuesRequest) (
+	*upiv1.PredictValuesResponse, error) {
+	var predictionErr *errors.TuringError // Measure execution time
+	defer metrics.Glob().MeasureDurationMs(
```

Comment:
I wonder if we can extract this to an external function? I feel it's a bit polluting the `PredictValues` code.